relay-server/src/services/buffer/envelope_repository/sqlite.rs

```diff
@@ -0,0 +1,662 @@
+        let envelope_store = SqliteEnvelopeStore::prepare(config).await?;
+        Ok(Self {
+            envelope_stacks: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        })
+    }
+
+    /// Creates a new [`SqliteEnvelopeRepository`] instance given a [`SqliteEnvelopeStore`].
+    pub fn new_with_store(config: &Config, envelope_store: SqliteEnvelopeStore) -> Self {
+        Self {
+            envelope_stacks: HashMap::new(),
+            envelope_store,
+            buffered_envelopes_size: 0,
+            disk_batch_size: config.spool_envelopes_stack_disk_batch_size(),
+            max_disk_size: config.spool_envelopes_max_disk_size(),
+        }
+    }
+
+    /// Initializes the envelope repository.
+    ///
+    /// Retrieves the project key pairs from the envelope store and creates
+    /// an initialization state.
+    pub async fn initialize(&mut self) -> InitializationState {
+        match self.envelope_store.project_key_pairs().await {
+            Ok(project_key_pairs) => {
+                self.initialize_empty_stacks(&project_key_pairs);
+                InitializationState::new(project_key_pairs)
+            }
+            Err(error) => {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to initialize the sqlite stack repository"
+                );
+                InitializationState::empty()
+            }
+        }
+    }
+
+    /// Pushes an envelope to the repository for the given project key pair.
+    ///
+    /// If the spool threshold is exceeded, it may trigger spooling to disk.
+    pub async fn push(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+        envelope: Box<Envelope>,
+    ) -> Result<(), SqliteEnvelopeRepositoryError> {
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        self.envelope_stacks
+            .entry(project_key_pair)
+            .or_default()
+            .envelopes
+            .push(envelope);
+
+        self.buffered_envelopes_size += 1;
+
+        Ok(())
+    }
+
+    /// Peeks at the next envelope for the given project key pair without removing it.
+    ///
+    /// If no envelope is in the buffer, it will be loaded from disk and a reference will be
+    /// returned.
+    pub async fn peek(
+        &mut self,
+        project_key_pair: ProjectKeyPair,
+    ) -> Result<Option<&Envelope>, SqliteEnvelopeRepositoryError> {
+        if self.memory_empty(project_key_pair) && self.should_check_disk(project_key_pair) {
+            let envelopes = self.unspool_from_disk(project_key_pair, 1).await?;
+            // If we have no envelopes in the buffer and no on disk, we can be safe removing the entry
+            // in the buffer.
+            if envelopes.is_empty() {
```

Comment:
This happens when we `pop` the last element and since we don't have a lookahead, we never know we have no envelopes left until we query and get new envelopes. So in the next `peek` called by the `EnvelopeBuffer` we see that we have memory empty and we still should check disk and this is when we get back the empty envelopes.